services/alerts/src/makeAlertsManager.ts

```diff
@@ -1,36 +1,39 @@
-import { Alert, AlertsManager, AlertsManagerAlert } from './types'
+import type { Alert, AlertRef, AlertsManager, AlertsMap } from './types'
 
-type SetAlertsFunction = React.Dispatch<
-    React.SetStateAction<AlertsManagerAlert[]>
->
-
-const createAlertManagerAlert = (
-    alert: Alert,
-    id: number,
-    remove: (id: number) => void
-): AlertsManagerAlert => ({
-    ...alert,
-    id,
-    remove: () => remove(id),
-})
+const toVisibleAlertsArray = (alertsMap: AlertsMap) =>
+    Array.from(alertsMap.values()).sort(
+        (a, b) => (a.id as number) - (b.id as number)
+    )
 
 export const makeAlertsManager = (
-    setAlerts: SetAlertsFunction
+    setAlerts: React.Dispatch<React.SetStateAction<Alert[]>>
 ): AlertsManager => {
+    const alertsMap: AlertsMap = new Map()
     let id = 0
 
-    const remove = (id: number) => {
-        setAlerts(alerts => alerts.filter(alert => alert.id !== id))
-    }
-    const add = (alert: Alert) => {
-        setAlerts(alerts => [
-            ...alerts,
-            createAlertManagerAlert(alert, ++id, remove),
-        ])
+    const add = (alert: Alert, alertRef: AlertRef) => {
+        if (!alert.id) {
+            id++
+        }
+
+        const alertId = alert.id || id
+        const alertsMapAlert = {
+            ...alert,
+            id: alertId,
```

Comment:
Consolidating as this is only used once, and changing to nullish coallescing operator to avoid potential issues with id of `0`
```suggestion
        const alertsMapAlert = {
            ...alert,
            id: alert.id ?? id,
```